### PR TITLE
Chain 420/arch withdrawals

### PR DESCRIPTION
**Pull request by bflood366**

### CODE REVIEW

File: backend/src/main/kotlin/xyz/funkybit/apps/api/services/ExchangeApiService.kt [hunk 1]
```diff
@@ -57,6 +61,8 @@ import xyz.funkybit.sequencer.proto.SequencerError
 import java.math.BigDecimal
 import java.math.BigInteger
 
+const val BASE_WITHDRAW_MESSAGE = "[funkybit] Please sign this message to authorize withdrawal of %s %s from the exchange to your wallet."
```
Comment (bflood366) [inline backend/src/main/kotlin/xyz/funkybit/apps/api/services/ExchangeApiService.kt:64]: This is the withdraw message to be signed. I followed other message but removed the sentence around not being charged a fee, since we will have withdrawal fees

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 2]
```diff
@@ -205,4 +320,98 @@ class ArchTransactionHandler(
     private fun submitToArch(transaction: BlockchainTransactionEntity, instruction: ArchNetworkRpc.Instruction) {
         transaction.markAsSubmitted(ArchUtils.signAndSendInstruction(instruction))
     }
+
+    private fun createNextWithdrawalBatch(programBitcoinAddress: BitcoinAddress, programPubkey: ArchNetworkRpc.Pubkey): Boolean {
+        var limit = 10
+
+        while (true) {
```
Comment (bflood366) [inline backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:327]: the concept here is to start with an upper limit of 10 withdrawals, and then it work backwards till it can find the number that fits within the MAX_TX_SIZE limit

File: backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt [hunk 3]
```diff
@@ -17,20 +20,31 @@ object UtxoSelectionService {
...
             unspentUtxos.filter { it.reservedBy == null },
             fee,
         )
     }
 
-    fun refreshUnspentUtxos(address: BitcoinAddress): List<UnspentUtxo> {
+    fun selectUtxosForProgram(amount: BigInteger, fee: BigInteger): List<UnspentUtxo> {
+        val address = DeployedSmartContractEntity.programBitcoinAddress()
+        val skipTxIds = listOf(ArchAccountEntity.findProgramAccount()!!.utxoId.txId())
+        val unspentUtxos = refreshUnspentUtxos(address, skipTxIds, skipMempool = true)
```
Comment (bflood366) [inline backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt:34]: the idea here is that for withdrawals we will only use confirmed utxos in the selection process.

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 4]
```diff
@@ -205,4 +320,98 @@ class ArchTransactionHandler(
...
+
+            val (instruction, selectedUtxos) = try {
+                ArchUtils.buildWithdrawBatchInstruction(
+                    programBitcoinAddress,
+                    programPubkey,
+                    sequencedWithdrawals,
+                )
+            } catch (e: BitcoinInsufficientFundsException) {
+                logger.warn(e) { "There are not enough funds to process a withdrawal batch" }
+                continue
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:350]: Will the loop just continue forever in this case? Or will it exit eventually on max age check? `batchMaxIntervalMs` is set to 1000 by default so I'm wondering what's the point if `continue` here? Are we expecting that the insufficient funds problem may get resolved within that interval?

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 5]
```diff
@@ -205,4 +320,98 @@ class ArchTransactionHandler(
...
+                    listOf(
+                        BalanceChange.Delta(
+                            withdrawalEntity.wallet.id.value,
+                            withdrawalEntity.symbol.id.value,
+                            withdrawalEntity.resolvedAmount().negate(),
+                        ),
+                    ),
+                    BalanceType.Exchange,
+                )
+                broadcasterNotifications.add(BroadcasterNotification.walletBalances(withdrawalEntity.wallet))
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:398]: Seems like we could just call `publishBroadcasterNotifications(listOf(BroadcasterNotification.walletBalances(withdrawalEntity.wallet)))` here

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/Withdrawal.kt [hunk 6]
```diff
@@ -160,6 +171,54 @@ class WithdrawalEntity(guid: EntityID<WithdrawalId>) : GUIDEntity<WithdrawalId>(
...
+        }
+
+        fun findSequencedArchWithdrawals(limit: Int, maxResponseSequence: Long?): List<SequencedArchWithdrawal> {
+            return WithdrawalTable
+                .join(ArchAccountTable, JoinType.INNER, WithdrawalTable.symbolGuid, ArchAccountTable.symbolGuid)
+                .join(ArchAccountBalanceIndexTable, JoinType.INNER, ArchAccountTable.guid, ArchAccountBalanceIndexTable.archAccountGuid, additionalConstraint = { ArchAccountBalanceIndexTable.walletGuid.eq(WithdrawalTable.walletGuid) })
+                .selectAll()
+                .where {
+                    WithdrawalTable.status.eq(WithdrawalStatus.Sequenced).and(ArchAccountBalanceIndexTable.status.eq(ArchAccountBalanceIndexStatus.Assigned)).let { query ->
+                        maxResponseSequence?.let { query.and(WithdrawalTable.responseSequence.less(it)) } ?: query
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/xyz/funkybit/core/model/db/Withdrawal.kt:208]: Could use `.andIfNotNull()` here instead of `?.let`:
```
WithdrawalTable.status.eq(WithdrawalStatus.Sequenced)
  .and(ArchAccountBalanceIndexTable.status.eq(ArchAccountBalanceIndexStatus.Assigned))
  .andIfNotNull(maxResponseSequence?.let {WithdrawalTable.responseSequence.less(it)})
```

File: backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt [hunk 7]
```diff
@@ -17,20 +20,31 @@ object UtxoSelectionService {
...
+        val unspentUtxos = refreshUnspentUtxos(address, listOf(), skipMempool = false)
         return inputsSelector.selectInputs(
             amount,
             unspentUtxos.filter { it.reservedBy == null },
             fee,
         )
     }
 
-    fun refreshUnspentUtxos(address: BitcoinAddress): List<UnspentUtxo> {
+    fun selectUtxosForProgram(amount: BigInteger, fee: BigInteger): List<UnspentUtxo> {
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt:31]: Not sure what does `forProgram` mean here

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 8]
```diff
@@ -16,60 +20,109 @@ import xyz.funkybit.core.model.db.BalanceEntity
...
-            var txInProgress: Boolean
+            var txInProgress = false
 
             while (true) {
                 try {
-                    txInProgress = transaction {
-                        processBalanceIndexBatch(programPubkey) ||
-                            processDepositBatch(programPubkey)
+                    transaction {
+                        txInProgress = processBalanceIndexBatch(programPubkey)
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:80]: `processBalanceIndexBatch`, `processDepositBatch` and `processWithdrawalBatch` are a little hard to follow because of the boolean return value. Also return type `Boolean` is kind of vague and without the context it's not clear what it means. How about replacing it with a DB query that would determine whether there are any transactions in progress? E.g:

```
processBalanceIndexBatch(programPubkey)
processDepositBatch(programPubkey)
processWithdrawalBatch(programBitcoinAddress, programPubkey)
Thread.sleep(if (transactionsInProgress()) activePollingIntervalInMs else inactivePollingIntervalInMs)
```

**Review by antonyshchenko — APPROVED** [review]
Great job! 👍

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 9]
```diff
@@ -183,6 +236,68 @@ class ArchTransactionHandler(
...
+        } else {
+            createNextWithdrawalBatch(programBitcoinAddress, programPubkey)
+        }
+    }
+
+    fun processWithdrawalsSettlingOnBitcoin(): Boolean {
+        val settlingWithdrawals = WithdrawalEntity.findSettlingOnBitcoin()
+        return if (settlingWithdrawals.isNotEmpty()) {
+            settlingWithdrawals.groupBy { it.blockchainTransactionGuid!! }.forEach { (_, withdrawals) ->
+                withdrawals.first().blockchainTransaction?.let { blockchainTransaction ->
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:275]: changed

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 10]
```diff
@@ -183,6 +236,68 @@ class ArchTransactionHandler(
         }
     }
 
+    private fun processWithdrawalBatch(programBitcoinAddress: BitcoinAddress, programPubkey: ArchNetworkRpc.Pubkey): Boolean {
+        val settlingWithdrawals = WithdrawalEntity.findSettlingOnArch()
+        logger.debug { "processWithdrawalBatch count = ${settlingWithdrawals.size}" }
+        return if (settlingWithdrawals.isNotEmpty()) {
+            settlingWithdrawals.first().archTransaction?.let { archTransaction ->
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:243]: changed

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt [hunk 11]
```diff
@@ -183,6 +236,68 @@ class ArchTransactionHandler(
...
+    }
+
+    fun processWithdrawalsSettlingOnBitcoin(): Boolean {
+        val settlingWithdrawals = WithdrawalEntity.findSettlingOnBitcoin()
+        return if (settlingWithdrawals.isNotEmpty()) {
+            settlingWithdrawals.groupBy { it.blockchainTransactionGuid!! }.forEach { (_, withdrawals) ->
+                withdrawals.first().blockchainTransaction?.let { blockchainTransaction ->
+                    val bitcoinTx = BitcoinClient.getRawTransaction(blockchainTransaction.txHash!!)
+                    if (bitcoinTx != null) {
+                        if ((bitcoinTx.confirmations ?: 0) >= numConfirmations) {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:278]: changed

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt (hunk 5 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:398]: yep, thx changed

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/Withdrawal.kt (hunk 6 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/model/db/Withdrawal.kt:208]: thx, changed

File: backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt (hunk 8 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/apps/ring/ArchTransactionHandler.kt:80]: I'm not sure having a different polling interval for active vs inactive is really necessary, so for now I just straight lined it as you suggested and just have one pollingInterval.

File: backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt (hunk 7 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/services/UtxoSelectionService.kt:31]: it specifically for UTXOs deposited to our arch program. I will change the name to something like 'selectUtxosDepositedToExchange` to make it a bit more descriptive

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V86_AddArchTransactionToWithdrawal.kt [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import xyz.funkybit.core.model.db.GUIDTable
+import xyz.funkybit.core.model.db.WithdrawalId
+
+@Suppress("ClassName")
+class V86_AddArchTransactionToWithdrawal : Migration() {
+
+    object V86_WithdrawalTable : GUIDTable<WithdrawalId>("withdrawal", ::WithdrawalId) {
+        val archTransactionGuid = reference(
+            "arch_tx_guid",
+            BlockchainTransactionTable,
```
Comment (bholzman) [inline backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V86_AddArchTransactionToWithdrawal.kt:16]: `V86_BlockchainTransactionTable`

Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V86_AddArchTransactionToWithdrawal.kt:16]: yep, thx
